### ISSUE TITLE
Ensure subprocess commands with stdin pipes are closed properly

### ISSUE THREAD

**Issue by technoweenie** [git-lfs/git-lfs#1874]
Opening this issue to continue the work started in #1860. I agree with @monitorjbl's asssesment in https://github.com/git-lfs/git-lfs/issues/1860#issuecomment-273022583:

> After writing the integration test for this fix, it looks like the root cause is that the stdout pipe isn't exhausted if the number of refs is very large. Since the code reads data through bufio.Scanner is buffered by some amount, the pipe is typically exhausted for most repos. However, if you have more than 900 refs in your repo, the amount of output from git for-each-ref is large enough that not all of it gets buffered.
>
> I wonder if this kind of issue is possibly the root cause of some of the other various hanging issues that folks have reported. There are several instances of commands just using Cmd.Wait() and not closing the associated pipe scattered around the codebase.

IMO, this seems like an unnecessary gotcha. I wonder if we can fix this in the `subprocess` package instead of changing the `defer` on every instance like #1861 does.

TODO:

* [ ] Merge #1861
* [ ] Backport #1861 for v1.5.x
* [ ] Fix other stdinpipe instances
* [ ] Backport those fixes for v1.5.x

**Comment by monitorjbl** [comment]
Go isn't my primary language, but it seems to me you could use a *try-with* pattern here. Something like this could work:

```
func ExecCmd(doWith func(io.WriteCloser, io.ReadCloser, io.ReadCloser), name string, arg ...string) error {
	cmd := subprocess.ExecCommand(name, arg...)

	//Open pipes
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return err
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return err
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return err
	}

	//Clean up resources when we're done
	defer func(){
		stdin.Close()
		stdout.Close()
		stderr.Close()
		cmd.Wait()
	}()

	// Start command and let doWith() play Mario with pipes
	if err := cmd.Start(); err != nil {
		return err
	}
	doWith(stdin, stdout, stderr)

	return nil
}

func main() {
	var outside string
	err := ExecCmd(func(stdin io.WriteCloser, stdout io.ReadCloser, stderr io.ReadCloser) {
		scanner := bufio.NewScanner(stdout)
		for scanner.Scan() {
			line := strings.TrimSpace(scanner.Text())
			fmt.Printf("%v\n", line)
		}
		outside = "we did it!"
	}, "git", "show-ref", "--heads", "--tags")

	//do stuff with any errors
	if err != nil {
		fmt.Printf("%v\n", err)
	} else {
		fmt.Printf("%v\n", outside)
	}
}

```

**Comment by technoweenie** [comment]
That could work, but I have something else in mind. The `subprocess` package already acts as a [really thin layer over the `os/exec` package](https://github.com/git-lfs/git-lfs/blob/f09e5d41ee2a1be5980efeae4ab829482a60739b/subprocess/subprocess_windows.go). I want to see if we can update it to return a wrapped `*exec.Cmd`, where calling `StdoutPipe()` returns a special `io.ReadCloser` that automatically closes when the command closes. This way, we're not creating unnecessary pipes, and callers don't have to worry about closing any opened pipes specifically.

**Comment by monitorjbl** [comment]
Sounds like a pretty low-resistance solution to me. Barely any code changes required, which is always nice!

**Comment by monitorjbl** [comment]
I took a stab at implementing what you suggested in #1861. Let me know if that's what you had in mind or if there is a cleaner way to do this.

**Comment by technoweenie** [comment]
Just merged #1861. Thanks again!

**Comment by marlazinger200** [comment]
1
